src/arr/trove/statistics.arr

```diff
@@ -244,11 +249,61 @@ fun matrix-based-multiple-regression(x_s_s :: List<List<Number>>, y_s :: List<Nu
   B_pred_fn
 end
 
-fun multiple-regression(x_s_s :: List<List<Number>>, y_s :: List<Number>) -> (List<Number> -> Number):
+fun js-based-multiple-regression(x_s_s :: List<List<Number>>, y_s :: List<Number>) -> (List<Number> -> Number):
   doc: "returns a predictor function given a list of list of independent inputs and the correspoinding list of outputs"
   MR.multiple-regression(x_s_s, y_s)
 end
 
+fun multiple-regression(x_s_s :: List<List<Number>>, y_s :: List<Number>) -> (List<Number> -> Number):
+  matrix-based-multiple-regression(x_s_s, y_s)
+end
+
+fun zip-num-lists(lol :: List<List<Number>>) -> List<List<Number>>:
+  num-cols = lol.length()
+  num-rows = lol.get(0).length()
+  for map(i from range(0, num-rows)):
+    lol.map(lam(l): l.get(i) end)
+  end
+end
+
+fun multiple-regression-coeffs(t :: Table, input-cols :: List<String>, output-col :: String):
+  doc: "returns a list of regression coefficients, given a table, a list of independent columns, and the corresponding output column"
+  # spy: t, input-cols, output-col end
+  x_s_s = zip-num-lists(input-cols.map(lam(col-name :: String): t.get-column(col-name) end))
+  # spy: x_s_s end
+  y_s = t.get-column(output-col)
+  # spy: y_s end
+  intercepts = MX.col-matrix.make(raw-array-of(1, y_s.length()))
+  # spy: intercepts end
+  mx_X = intercepts.augment(MX.lists-to-matrix(x_s_s))
+  # spy: mx_X end
+  mx_Y = MX.col-matrix.make(raw-array-from-list(y_s))
+  # spy: mx_Y end
+  MX.mtx-to-list(MX.mtx-least-squares-solve(mx_X, mx_Y))
+end
+
+fun multiple-regression-table(t :: Table, input-cols :: List<String>, output-col :: String):
+  doc: "returns a regression coefficient table, given a table, a list of independent columns, and the corresponding output column"
+  B = multiple-regression-coeffs(t, input-cols, output-col)
+  coeff-names = input-cols.push("constant")
+  [table-from-columns: {"coefficient-name"; coeff-names}, {"coefficient-value"; B}]
+end
+
+fun multiple-regression-fun(t :: Table, input-cols :: List<String>, output-col :: String) -> (List<Number> -> Number):
```

Comment:
This function can't possibly be needed as written; don't we have `multiple-regression` already?  This looks like copy-pasted code, and fiddly code at that, so we should deduplicate it.

Reply:
The difference is that the current trio of functions take a table, and column names as arguments. The Bootstrap group decided that all three outputs (viz., list of coefficients, table of coefficients, and predictor function) were needed, and it wasn't clear which one (or perhaps two) should be treated as basic.

Reply:
That's not my point.  We already have https://github.com/brownplt/pyret-lang/blob/horizon/src/arr/trove/statistics.arr#L236-L244 that defines `B_pred_fun`; don't duplicate that code.

Reply:
It's not a duplicate. `multiple-regression-fun` calls `multiple-regression-coeffs` to get a list of coefficients, which it calls `B`.  The `B_pred_fn` then uses `fold2` to do a dot-product its argument list with `B`.  It avoids having to convert the argument list to a vector.

In the original `multiple-regression`, we get a _vector_ of coefficients, which it calls `B`. Its `B_pred_fn` converts its argument list into a vector and then performs a vector dot product with the vector `B`.  

The only commonality between the two `B_pred_fn`s is that they check the length of their incoming list the same way to decide if an exception needs to be thrown. Once past that check, the body is different.

One could write an external function that is called in both places (with `B` as argument) to create `B_pred_fn`, but even so, one would have to do an extra list/vector casting in one of the two functions, so that the `B`s are the same type.